org/mozilla/jss/provider/javax/crypto/JSSMacSpi.java

```diff
@@ -21,9 +23,9 @@
 class JSSMacSpi extends javax.crypto.MacSpi {
 
     private JSSMessageDigest digest=null;
-    private HMACAlgorithm alg;
+    private DigestAlgorithm alg;
```

Comment:
@edewata This isn't quite right I now realize (I thought it wasn't, then though it was, now realize it isn't). 

DigestAlgorithm defines things like SHA-1, SHA-2, &c. 

HMACAlgorithm is a subclass which defines things like HMAC/SHA-1, HMAC-SHA-2-256, &c. These all require keys though, unlike digests (though, some digests are keyed, but none we define are).  

CMAC (the algorithm) isn't actually based off of a digest (hash algorithm), but is instead based off a (keyed) cipher. 

So technically, we should have a class Structure like:

```
Algorithm
  DigestAlgorithm
  MACAlgorithm
    CMACAlgorithm
    HMACAlgorithm
```

But currently we have the following class structure:

```
Algorithm
  DigestAlgorithm
    CMACAlgorithm
    HMACAlgorithm
```

Lastly, these are protected so end-users won't directly be constructing them. I see the following actions we could take:

 1. Be happy how it is.
 2. Add an explicit check for HMAC or CMAC subclass. 
 3. (A breaking change for anyone using HMACAlgorithm) Fix the class structure. 
 4. Make CMAC inherit from Algorithm or a new MACAlgorithm, not DigestAlgorithm, checking either for HMACAlgorithm or MACAlgorithm.

Note that any checking that we would do here is limited to developers adding new HMAC/CMAC/*MAC implementations and making their lives easier. 


Thoughts?

Reply:
(UMAC is the only other MAC implementation I could see someone wanting to add, but NSS doesn't yet support it).

Reply:
I'm OK with #3 for a new JSS minor version (4.7?). If we want to keep the current minor version we can use any other options to provide the new functionality without introducing incompatible changes, but consider that as a temporary solution until we fix the structure. As long as we document the changes clearly it should be OK to require people to change their code when upgrading JSS.

Reply:
How about we add that as a second patchset after this merges, since it involves changing both HMAC and CMAC, and we can do additional simplifications internally?

Reply:
Sure, thanks!